builder/amazon/common/step_run_spot_instance.go

```diff
@@ -306,25 +303,19 @@ func (s *StepRunSpotInstance) Run(ctx context.Context, state multistep.StateBag)
 		return err
 	})
 
-	// Create the request for the spot instance.
-	if err != nil {
+	// We can end up with errors because one of the allowed availability
+	// zones doesn't have one of the allowed instance types; as long as
+	// an instance is launched, these errors aren't important.
+	if err != nil && len(createOutput.Instances) == 0 {
 		if createOutput.FleetId != nil {
 			err = fmt.Errorf("Error waiting for fleet request (%s): %s", *createOutput.FleetId, err)
-		} else {
-			err = fmt.Errorf("Error waiting for fleet request: %s", err)
 		}
-		// We can end up with errors because one of the allowed availability
-		// zones doesn't have one of the allowed instance types; as long as
-		// an instance is launched, these errors aren't important.
 		if len(createOutput.Errors) > 0 {
 			errString := fmt.Sprintf("Error waiting for fleet request (%s) to become ready:", *createOutput.FleetId)
 			for _, outErr := range createOutput.Errors {
 				errString = errString + aws.StringValue(outErr.ErrorMessage)
 			}
 			err = fmt.Errorf(errString)
-			state.Put("error", err)
-			ui.Error(err.Error())
-			return multistep.ActionHalt
 		}
 		state.Put("error", err)
 		ui.Error(err.Error())
```

Comment:
ah 🤦  I should have seen that in my first pass, but should this error handling be in the retry logic too ? 

So :
* If an instance was created don't retry and don't check the error
* retry if the error is retryable

Reply:
Yeah, I think it might make sense to move it inside the retry 🤔

Reply:
Tried moving inside the retry logic and it started to become a bit messy. Had to check for nil values and error more than the current code

Reply:
Ah, noes, but I think it's important because then we might create multiple instances from the retry.

Reply:
But it already worked the way it is. The reported tested and is not creating multiple instances anymore. The retry will only retry with the iamInstanceProfile error which means it won't be possible to create any instance without it.

Reply:
Got it, yes !